src/util/bytes.cpp

```diff
@@ -90,4 +93,57 @@ std::string formatByteArrayToIntString(const std::vector<uint8_t>& bytes)
 
     return ss.str();
 }
+
+// This function is called in a tight loop over large regions of data so
+// make sure it stays efficient.
+std::vector<std::pair<uint32_t, uint32_t>> diffArrayRegions(
+  std::span<const uint8_t> a,
+  std::span<const uint8_t> b)
+{
+    PROF_START(ByteArrayDiff)
+    std::vector<std::pair<uint32_t, uint32_t>> regions;
+
+    // Iterate through diffs and work out start and finish offsets of each dirty
+    // region
+    uint32_t diffStart = 0;
+    bool diffInProgress = false;
+    for (uint32_t i = 0; i < a.size(); i++) {
+        bool dirty = a.data()[i] != b.data()[i];
+        if (dirty && !diffInProgress) {
+            // Starts at this byte
+            diffInProgress = true;
+            diffStart = i;
+        } else if (!dirty && diffInProgress) {
+            // Finished on byte before
+            diffInProgress = false;
+            regions.emplace_back(diffStart, i - diffStart);
+        }
+    }
+
+    // If we finish with a diff in progress, add it
+    if (diffInProgress) {
+        regions.emplace_back(diffStart, a.size() - diffStart);
+    }
+
+    PROF_END(ByteArrayDiff)
+    return regions;
+}
+
+std::vector<bool> diffArrays(std::span<const uint8_t> a,
+                             std::span<const uint8_t> b)
+{
+    if (a.size() != b.size()) {
+        SPDLOG_ERROR(
+          "Cannot diff arrays of different sizes {} != {}", a.size(), b.size());
+        throw std::runtime_error("Cannot diff arrays of different sizes");
+    }
+
+    std::vector<bool> diffs(a.size(), false);
+    for (int i = 0; i < a.size(); i++) {
+        diffs[i] = a.data()[i] != b.data()[i];
```

Comment:
This might be *very* slow due to the `vector<bool>` specialization, just swapping the type to uint8_t (at the cost of 8x more memory) makes it 11x faster: <https://quick-bench.com/q/d8INA76hIDM3m4gfC4xJKI4HcQM>